# **E-commerce Analysis Project**



  ## **Case Study**

  This project focuses on leveraging data analytics techniques to derive actionable insights from an e-commerce sales dataset. The analysis aims to enhance business decision-making by identifying sales trends, segmenting customers, evaluating product performance, and predicting future sales patterns. These insights can be used by businesses to optimize inventory, tailor marketing strategies, and improve customer engagement.

## **Data Overview**

The dataset contains transactional records of an e-commerce business, providing information on invoices, products, customers, and sales across multiple countries.

- InvoiceNo: Unique identifier for each transaction/invoice.
- StockCode: Unique code for each product.
- Description: Textual description of the product.
- Quantity: Number of products purchased (can be negative for returns).
- InvoiceDate: Timestamp of the transaction.
- UnitPrice: Price of a single unit of the product.
- CustomerID: Unique identifier for the customer.
- Country: The country where the transaction occurred.


# **Import Important Libraries**

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.express as px
import time

# **Read Dataset**

In [ ]:
data = pd.read_csv("data.csv", encoding="latin1")

# **Exploring Data**

In [ ]:
data.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,12/1/2010 8:26,7.65,17850.0,United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,12/1/2010 8:26,4.25,17850.0,United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,12/1/2010 8:28,1.85,17850.0,United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,12/1/2010 8:28,1.85,17850.0,United Kingdom
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,12/1/2010 8:34,1.69,13047.0,United Kingdom


In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [ ]:
data.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


In [ ]:
print(data.dtypes)

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object


# **Task1: Apply Necessary Preprocessing**
> * **Check for Missing Values and Duplicates and Handle them**
> * **Convert InvoiceDate Column to DateTime Type**
> * **Check if Quantity Column or UnitPrice Column Contains Negative Values and Remove them if Found**
> * **Create a TotalPrice column**

**Check for Missing Values and Duplicates**

In [ ]:
data.isnull().sum()

,0
InvoiceNo,0
StockCode,0
Description,1454
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,135080
Country,0


In [ ]:
# Check for duplicate
data.duplicated().sum()

5268

**Drop Nulls and Duplicates**

In [ ]:
cleaned_data = data.dropna(axis = 0, inplace = True)

In [ ]:
cleaned_data = data.drop_duplicates()

**Convert InvoiceDate Column to DateTime Type**

In [ ]:
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'], errors='coerce')

<ipython-input-11-c4730cbcb831>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'], errors='coerce')


**Check if Quantity Column or UnitPrice Column Contains Negative Values and Remove them if Found**

In [ ]:
# Check if there are any rows with negative Quantity or UnitPrice
negative_rows = data[(data['Quantity'] < 0) | (data['UnitPrice'] < 0)]

# Display the count of rows with negative values
print(f"Number of rows with negative Quantity or UnitPrice: {negative_rows.shape[0]}")

# Drop rows with negative Quantity or UnitPrice
cleaned_data = cleaned_data.drop(cleaned_data[(cleaned_data['Quantity'] < 0) | (cleaned_data['UnitPrice'] < 0)].index)

Number of rows with negative Quantity or UnitPrice: 8905


**Create a TotalPrice column**

In [ ]:
# Create TotalSales column
cleaned_data['Totalprice'] = data['Quantity'] * data['UnitPrice']

# **Task2: Perform Exploratory Data Analysis**
> * **View Statistical Summary for Numerical Columns**
> * **Identify Top 10 Selling Products by Quantity**
> * **Calculate the Total Revenue**
> * **Calculate the Number of Transactions**

In [ ]:
cleaned_data.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID,Totalprice
count,392732.000000,392732,392732.000000,392732.000000,392732.000000
mean,13.153718,2011-07-10 19:15:24.576301568,3.125596,15287.734822,22.629195
min,1.000000,2010-12-01 08:26:00,0.000000,12346.000000,0.000000
25%,2.000000,2011-04-07 11:12:00,1.250000,13955.000000,4.950000
50%,6.000000,2011-07-31 12:02:00,1.950000,15150.000000,12.390000
75%,12.000000,2011-10-20 12:53:00,3.750000,16791.000000,19.800000
max,80995.000000,2011-12-09 12:50:00,8142.750000,18287.000000,168469.600000
std,181.588420,NaN,22.240725,1713.567773,311.083465


In [ ]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392732 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    392732 non-null  object        
 1   StockCode    392732 non-null  object        
 2   Description  392732 non-null  object        
 3   Quantity     392732 non-null  int64         
 4   InvoiceDate  392732 non-null  datetime64[ns]
 5   UnitPrice    392732 non-null  float64       
 6   CustomerID   392732 non-null  float64       
 7   Country      392732 non-null  object        
 8   Totalprice   392732 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(4)
memory usage: 30.0+ MB


In [ ]:
cleaned_data

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Totalprice
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
...,...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France,10.20
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France,12.60
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France,16.60
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France,16.60


**Identify Top 10 Selling Products by Quantity**

In [ ]:
# Group by StockCode and sum the quantities
# Note: we choosed to group by StockCode rather than Description as the Description has no clear pattern,
# there can be same product but written in 2 different ways, so the stock code will ensure unique identification and accurate results
top_products = cleaned_data.groupby('StockCode')['Quantity'].sum().sort_values(ascending=False).head(10).reset_index()

# Display the top 10 selling products
print("Top 10 Selling Products by Quantity:")
top_products

Top 10 Selling Products by Quantity:


,StockCode,Quantity
0,23843,80995
1,23166,77916
2,84077,54319
3,22197,49160
4,85099B,46078
5,85123A,36763
6,84879,35263
7,21212,33670
8,23084,27153
9,22492,26076


In [ ]:
# Function to extract the top 10 products' Description from its StockCode
def get_product_description(stockcode):

    product_desc = cleaned_data[cleaned_data['StockCode'] == stockcode].iloc[0]['Description']
    return product_desc

top_products['Description'] = top_products['StockCode'].apply(get_product_description)
print(top_products)

  StockCode  Quantity                         Description
0     23843     80995         PAPER CRAFT , LITTLE BIRDIE
1     23166     77916      MEDIUM CERAMIC TOP STORAGE JAR
2     84077     54319   WORLD WAR 2 GLIDERS ASSTD DESIGNS
3     22197     49160                SMALL POPCORN HOLDER
4    85099B     46078             JUMBO BAG RED RETROSPOT
5    85123A     36763  WHITE HANGING HEART T-LIGHT HOLDER
6     84879     35263       ASSORTED COLOUR BIRD ORNAMENT
7     21212     33670     PACK OF 72 RETROSPOT CAKE CASES
8     23084     27153                  RABBIT NIGHT LIGHT
9     22492     26076             MINI PAINT SET VINTAGE 


In [ ]:
# Visualize top-selling products
fig = px.bar(top_products.head(10), x='Description', y='Quantity', title='Top 10-Selling Products')
fig.update_layout(xaxis_title='Product Name', yaxis_title='Quantity Sold')

fig.show()

**Calculate Total Revenue**

In [ ]:
total_revenue = cleaned_data['Totalprice'].sum()
print(f"Total Revenue: ${total_revenue:.2f}")

Total Revenue: $8887208.89


**Calculate the Number of Transactions**

In [ ]:
# Calculate the total number of transactions
total_transactions = cleaned_data['InvoiceNo'].nunique()  # Assuming unique InvoiceNo means unique transactions
print(f"Total Transactions: {total_transactions}")

Total Transactions: 18536


# **Task3: Time Series Analysis "Monthly sales trends"**
###  Here we are gonna identify the monthly trends.

In [ ]:
# Extract month from InvoiceDate
cleaned_data['Month'] = cleaned_data['InvoiceDate'].dt.month
cleaned_data.head(5)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Totalprice,Month
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30,12
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,12
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00,12
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,12
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,12


In [ ]:
cleaned_data['InvoiceDate'].dt.year.value_counts()

,count
InvoiceDate,
2011,367059
2010,25673


In [ ]:
# Calculate monthly sales using sum of quantities sold for each month
monthly_sales = cleaned_data.groupby('Month')['Quantity'].sum().reset_index()
monthly_sales.columns = ['Month', 'TotalSales']
monthly_sales

,Month,TotalSales
0,1,348522
1,2,265043
2,3,347623
3,4,291369
4,5,372948
5,6,363014
6,7,367372
7,8,398190
8,9,543654
9,10,591551


In [ ]:
# Visualize monthly sales (Time Series Analysis)
month_names = [
    'January', 'February', 'March', 'April', 'May', 'June',
    'July', 'August', 'September', 'October', 'November', 'December'
]
monthly_sales['MonthName'] = monthly_sales['Month'].map(lambda x: month_names[x - 1])

fig = px.line(
    monthly_sales,
    x='MonthName',
    y='TotalSales',
    title='Monthly Sales Trends',
    labels={'X': 'Month', 'Y': 'Total Sales'},
    markers=True
)

fig.update_traces(marker=dict(size=8), line=dict(color='red', width=2))
fig.update_layout(title_font_size=18)
fig.show()

## Insights gained from the the monthly trend analysis are:
- The Month with the highest sale is: November.
- The Month with the lowest sales is: February.

# **Task4: RFM Analysis & Customer Segmentation**

### Segment customers based on:


*   Recency: Days since last purchase.
*   Frequncy: Number of purchases.
*   Monetary: Total spent amount.







In [ ]:
# Convert CustomerID's datatype to string and fix its format
# Regex expression purpose: remove .0 from customerID
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].astype(str).str.replace(r'\.0$', '', regex=True)
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392732 entries, 0 to 541908
Data columns (total 10 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    392732 non-null  object        
 1   StockCode    392732 non-null  object        
 2   Description  392732 non-null  object        
 3   Quantity     392732 non-null  int64         
 4   InvoiceDate  392732 non-null  datetime64[ns]
 5   UnitPrice    392732 non-null  float64       
 6   CustomerID   392732 non-null  object        
 7   Country      392732 non-null  object        
 8   Totalprice   392732 non-null  float64       
 9   Month        392732 non-null  int32         
dtypes: datetime64[ns](1), float64(2), int32(1), int64(1), object(5)
memory usage: 31.5+ MB


In [ ]:
# Explore number of customers
print(cleaned_data['CustomerID'].nunique())

4339


**Recency**

In [ ]:
# Calculate number of days since last purchase per customer (Recency)
# Group data by CustomerID and calculate the difference between each customer's last purchase day and a reference day (last day in dataset)
cleaned_data['Recency'] = cleaned_data.groupby('CustomerID')['InvoiceDate'].transform(lambda x: (cleaned_data['InvoiceDate'].max() - x.max()).days)

**Frequency**

In [ ]:
# Calculate number of purchases per customer (Frequency)
# Group data by CustomerID and find number of unique Invoices for each customer
cleaned_data['Num_of_Purchases'] = cleaned_data.groupby('CustomerID')['InvoiceNo'].transform('nunique')

**Monetary**

In [ ]:
# Clculate total spend per customer (Monetary)
# Group data by CustomerID and calculate sum of Totalprice for each item a customer purchased
cleaned_data['Total_Spend'] = cleaned_data.groupby('CustomerID')['Totalprice'].transform('sum')

**Customer Segmentation**

In [ ]:
# Create new dataset for customer segmentation
# Group data by CustomerID and take only one value to represent each matric of (RFM)
customer_segments = cleaned_data.groupby('CustomerID').agg({
    'Recency': 'min',
    'Num_of_Purchases': 'min',
    'Total_Spend': 'min'
}).reset_index()
customer_segments

,CustomerID,Recency,Num_of_Purchases,Total_Spend
0,12346,325,1,77183.60
1,12347,1,7,4310.00
2,12348,74,4,1797.24
3,12349,18,1,1757.55
4,12350,309,1,334.40
...,...,...,...,...
4334,18280,277,1,180.60
4335,18281,180,1,80.82
4336,18282,7,2,178.05
4337,18283,3,16,2045.53


In [ ]:
# Explore summary statistics of RFM matrices to define a suitable segmentation strategy
customer_segments.describe()

,Recency,Num_of_Purchases,Total_Spend
count,4339.000000,4339.000000,4339.000000
mean,91.518322,4.271952,2048.215924
std,100.009747,7.705493,8984.248352
min,0.000000,1.000000,0.000000
25%,17.000000,1.000000,306.455000
50%,50.000000,2.000000,668.560000
75%,141.000000,5.000000,1660.315000
max,373.000000,210.000000,280206.020000


In [ ]:
# Visualiza Distibution of RFM matrices to explore the shape of data
fig = make_subplots(rows=1, cols=3, subplot_titles=['Recency', 'Frequency', 'Monetary'])

# Add Recency distribution
fig.add_trace(
    go.Histogram(
        x=customer_segments['Recency'],
        nbinsx=20,
        name='Recency',
        marker=dict(color='skyblue'),
    ),
    row=1, col=1
)

# Add Frequency distribution
fig.add_trace(
    go.Histogram(
        x=customer_segments['Num_of_Purchases'],
        nbinsx=20,
        name='Frequency',
        marker=dict(color='orange'),
    ),
    row=1, col=2
)

# Add Monetary distribution
fig.add_trace(
    go.Histogram(
        x=customer_segments['Total_Spend'],
        nbinsx=20,
        name='Monetary',
        marker=dict(color='green'),
    ),
    row=1, col=3
)

fig.update_layout(
    title_text="Distribution of RFM Matrices",
    title_font_size=16,
    showlegend=False
)

fig.show()

# **Observations**



1.   Recency:

        *   Values renge between 0 to 373 days.
        *   50% of customers made a purchase 50 days ago.
        *   Large variance indicates a wide spread in values.
        *   75% of customers have Recency ≤ 141 days, meaning a large portion   of customers made relatively recent purchases.

2.   Number of Purchases:

        *   Values renge between 1 to 210 purchases.
        *   50% of customers made 2 purchases.
        *   75% of customers made ≤ 5 purchases, indicating that most customers are low-frequency buyers.


3.   Total Spend:

        *   large variance ($8,984), suggesting the presence of extreme outliers.
        *   Maximum ($280,206.02) is an extreme outlier.
        *   75% of customers spent ≤ $1,660.31, meaning most customers are low to medium spenders.


**Note:** From Visualizations it's clear that **all RFM metrices are skwed and have outliers** (especially Frequency and Monetary) which is an important observation we need to consider during segmentation.




# **Segmentation Strategy (Scoring-based Method)**

**Steps:**



1.   Devide each RFM metric into 3 bins (High = 3, Medium = 2, Low = 1) and specify intervals based on percentiles as this method work well with skewed data and outliers.

      *   Recency: High = 0:17, Medium = 18:141, Low > 141
      *   Frequency: High ≥ 5, Medium = 2:4, Low = 1
      *   Monetary: High ≥ $1,660.31, Medium = $306.46:$1,660.31, Low ≤ $306.46


2.   Calculate the total RFM score (range between 3 and 9).

3.   Segment Customers based on total RFM score:

      *   High-Value --> Greater than or equal 8.
      *   Medium-Value --> between 5 and 7.
      *   Low-Value --> less than or equal 4.


In [ ]:
# Define Ranges
recency_bins = [0, 17, 141, float('inf')]
frequency_bins = [1, 2, 5, float('inf')]
monetary_bins = [0, 306.46, 1660.31, float('inf')]

# Assign scores to each RFM metric usinf binning
customer_segments['R_score'] = pd.cut(customer_segments['Recency'], bins=recency_bins, labels=[3, 2, 1], right=False)
customer_segments['F_score'] = pd.cut(customer_segments['Num_of_Purchases'], bins=frequency_bins, labels=[1, 2, 3], right=False)
customer_segments['M_score'] = pd.cut(customer_segments['Total_Spend'], bins=monetary_bins, labels=[1, 2, 3], right=False)

In [ ]:
# Convert scores to integers
customer_segments['R_score'] = customer_segments['R_score'].astype(int)
customer_segments['F_score'] = customer_segments['F_score'].astype(int)
customer_segments['M_score'] = customer_segments['M_score'].astype(int)

# Calculate total RFM score
customer_segments['RFM_score'] = customer_segments['R_score'] + customer_segments['F_score'] + customer_segments['M_score']

In [ ]:
# Assign segments based on total RFM score
def assign_segment(score):
    if score >= 8:
        return 'High-Value'
    elif 5 <= score <= 7:
        return 'Medium-Value'
    else:
        return 'Low-Value'

customer_segments['Segment'] = customer_segments['RFM_score'].apply(assign_segment)
customer_segments

,CustomerID,Recency,Num_of_Purchases,Total_Spend,R_score,F_score,M_score,RFM_score,Segment
0,12346,325,1,77183.60,1,1,3,5,Medium-Value
1,12347,1,7,4310.00,3,3,3,9,High-Value
2,12348,74,4,1797.24,2,2,3,7,Medium-Value
3,12349,18,1,1757.55,2,1,3,6,Medium-Value
4,12350,309,1,334.40,1,1,2,4,Low-Value
...,...,...,...,...,...,...,...,...,...
4334,18280,277,1,180.60,1,1,1,3,Low-Value
4335,18281,180,1,80.82,1,1,1,3,Low-Value
4336,18282,7,2,178.05,3,2,1,6,Medium-Value
4337,18283,3,16,2045.53,3,3,3,9,High-Value


# **Visualize Customer Segments & Relationship Between RFM Metrices**

In [ ]:
fig = px.histogram(customer_segments,
                   x='Segment',
                   title="Customer Distribution by Segment",
                   labels={'Segment': 'Segment'},
                   color='Segment')

fig.update_layout(
    xaxis_title="Segment",
    yaxis_title="Number of Customers",
    template="plotly_white"
)

fig.show()

In [ ]:
fig = px.scatter(
    data_frame=customer_segments,
    x='Recency',
    y='Num_of_Purchases',
    color='Segment',
    size='Total_Spend',
    size_max=100,
    title='RFM Metrices Analysis',
    labels={
        'Recency': 'Recency (Days)',
        'Num_of_Purchases': 'Number of Purchases',
        'Segment': 'Customer Segment',
        'Total_Spend': 'Total Spend'
    }
)

fig.update_layout(
    legend_title=dict(text='Customer Segment'),
    xaxis=dict(title='Recency (Days)'),
    yaxis=dict(title='Number of Purchases')
)

# **Insights**


- High-value customers generally cluster with higher purchase frequencies and Low-value customers tend to have a much smaller number of purchases.
- High-value customers often have large bubbles (represents total spend), indicating significant contributions to revenue.
- High-value customers are driving the majority of revenue. Monitoring and enhancing their experience will have the most impact on overall profitability.
- Medium-value customers have potential to be upsold or cross-sold into higher-value customers. We could target them with tailored marketing campaigns.
- Low-value customers who have high recency might be responsive to re-engagement campaigns. For example, discounts or follow-up communications can bring them back.

# **Task5: Product Category Analysis**
- Analysis of sales and revenue by product category.
- Visualization of the top categories by revenue.

In [ ]:
# Explore product Description
cleaned_data['Description'].value_counts()

,count
Description,
WHITE HANGING HEART T-LIGHT HOLDER,2016
REGENCY CAKESTAND 3 TIER,1714
JUMBO BAG RED RETROSPOT,1615
ASSORTED COLOUR BIRD ORNAMENT,1395
PARTY BUNTING,1390
...,...
GREEN WITH METAL BAG CHARM,1
WHITE WITH METAL BAG CHARM,1
BLUE/NAT SHELL NECKLACE W PENDANT,1


In [ ]:
# Use a list of keywords to help identify the category of a product based on description.
category_keywords = {
    "Fine Art, Collectibles, Crafts & Sewing": ["ART", "KNICK KNACK", "METAL PEARS", "WOOD", "FLOWER", "GLOBE", "PHOTO CUBE", "CLAY", "KNITTING", "KNITTED", "CRAFT", "WOOD STAMP", "GARLAND", "SKULLS", "TREASURE", "DRAWING", "WOOD BOX", "WOODEN", "MAGNETS", "TAGS", "BADGES", "STAMP", "SCISSOR", "OWLS", "RECYCLED", "STICKERS", "GOLD", "WRAP", "BABUSHKA", "TAPE", "HAND", "WOOD BIRD", "WICKER", "WOOD STAMP", "CERAMIC", "FLYING DUCKS", "BABUSHKA NESTING BOXES", "TRINKET BOX", "HEART", "ORNAMENT", "GIFT TAPE", "SEWING", "BLOCK LETTERS", "CHALKBOARD", "BAUBLES"],
    "Decorations" : ["DECORATIONS", "DECORATION", "DECORATIVE", "VINTAGE", "ANTIQUE", "FRENCH"],
    "Lightening" : ["LIGHT", "LIGHTS", "LANTERN", "LAMP", "DISCO BALL", "BULB"],
    "Kitchenware" : ["BEAKERS", "COLANDER", "PLACEMAT", "OVEN", "ORANGE", "LOLLY", "OIL", "SPOON", "FORK", "KNIFE", "CHOPPING BOARD", "BEAKER", "NAPKINS", "BISCUIT", "MAKERS", "TOAST", "TEAPOT", "TEA", "TEACUP", "TRAY", "TRAYS", "COFFEE", "CHEESE", "CHOPSTICKS", "CAKE", "COOKIE", "STRAWS", "SUGAR", "CASES", "CUP","CUTLERY", "DINER", "DINNER", "SNACK", "KITCHEN", "BAKING", "FOOD COVER", "LUNCH", "FRIDGE MAGNETS", "BOWLS", "RECIPE", "GLASS", "SPICE", "PAPER PLATES", "BOTTLE", "BOTTLES", "CRACKERS", "JAR", "JUG", "JAM", "MUG", "BOWL", "PAN", "POT", "DISH", "PLATTER", "PLATE", "BREAKFAST", "BREAD", "EGG", "POPCORN HOLDER", "OVEN GLOVE", "PLACEMATS", "COASTERS", "MATCHES", "ICE CREAM"],
    "Home & Living & Garden" : ["HOOK", "STORAGE BOX", "BUCKET", "IRONING", "BOX", "BATH", "WINDMILL", "INCENSE", "ALBUM", "TIN", "PLANT", "GROW", "PLANTS", "PORTRAIT FRAME", "PHOTO FRAME", "FRAME", "MARKER", "GARDEN", "PICTURE FRAME", "RACK", "VASE", "DRAWER CABINET", "SOAP", "CABINET", "SIGN", "CLOCK", "CANDLES", "CANDLE", "CANDLE STICK", "DOORMAT", "BRUSH", "DOORSTOP", "TV DINNER TRAY", "HOME", "HANGER", "CUPBOARD", "T-LIGHT HOLDER", "MIRROR", "MIRRORED", "WALL ART", "DRAWER KNOB", "PHOTO FRAME", "TISSUE BOX", "TREASURE CHEST", "DRAWERS", "DRAWER"],
    "Seasonal & Holiday" :["CHRISTMAS", "CONFETTI", "STICKY BUNNIES", "EASTER", "SANTA", "WREATH", "NOEL", "VALENTINE", "JINGLE BELL", "HALLOWEEN", "WINTER HOLIDAY WREATH", "HOLLY", "BELLS"],
    "Textiles & Soft Furnishings" : ["BUNTING", "APRON", "CUSHION", "QUILT", "HAND WARMER", "TOWEL", "PAD", "CUSHION COVER", "RIBBON", "RIBBONS"],
    "Toys & Games" : ["GAME", "PONY", "SANDCASTLE", "CREEPY CRAWLIES", "FLYING DISC", "SPACE","ANIMAL", "TRADITIONAL CRAYONS", "SPINNING", "LUDO", "NAUGHTS & CROSSES", "FUN", "BALLOONS", "BALLOON", "TEDDY BEAR", "FUNKY MONKEY", "BLOCK WORD", "DINOSAUR", "PLAYHOUSE", "PLAYING CARDS", "SOLDIER SKITTLES", "SKIPPING ROPE", "SNAKES & LADDERS", "FELTCRAFT", "KIDS", "SPACEBOY", "CHILDRENS", "Child’s", "DOLLY GIRL", "TOY", "JIGSAW", "PLASTERS IN TIN", "CIRCUS PARADE", "HELICOPTER", "DOMINOES"],
    "Paper Products" : ["JOURNAL", "SHEET", "ENVELOPE", "CARD", "CALENDAR", "PAPER", "LEAVES", "NAPKIN", "CARDS", "TISSUES", "TISSUE"],
    "Office Supplies" : ["PASSPORT", "BLACK BOARD", "PHOTO CLIP", "RULER", "PENCILS", "PENS", "PEN", "PENCIL", "ERASERS", "LETTER HOLDER", "CARDHOLDER", "NOTEBOOK", "NOTEBOOKS", "SKETCHBOOK", "WASTEPAPER BIN", "Manual", "CHALK", "CHALK STICKS"],
    "Luggage & Travel Gear": ["AID", "LUGGAGE", "JUMPO BAG", "RAIN PONCHO"],
    "Clothing, Shoes, Accessories, Jewelry & Beauty" : ["SHOES", "FUR", "BAG", "SLIPPER", "TIARAS", "BANGLE", "SOMBRERO", "SOCK", "CLIPS", "BACKPACK", "GLOVES", "UMBRELLA", "HAIR", "SILK FAN", "HANDBAG", "GIRL", "PURSE","EDWARDIAN", "HAT", "GIRLY", "BRACELET", "EARRINGS", "NECKLACE", "RING", "BROOCH", "HAIR SLIDES", "SILVER", "HAIR TIE", "EDWARDIAN PARASOL", "JEWELLERY", "LIP GLOSS"],
    "Musical Instruments" : ["HARMONICA", "RECORD FRAME", "RECORD COVER FRAME"],
    "Books" : ["BOOKS", "BOOK"],
    "Tools & Home Improvement" : ["KIT", "REPAIR", "SET", "METAL BOX", "SCREWDRIVER"],
    "Sports & Outdoors" : ["BICYCLE", "TORCH", "B'FLY"],
    "Baby Supplies" : ["BABY", "CHILDREN'S APRON", "CHILDREN"],
    "Automotive Parts & Accessories": ["KEY", "BUTTON BOX", "CAR"],
    "Appliances" : ["HONEYCOMB FAN" ,"HEADPHONES"]
}

cleaned_data['Product Category'] = "Other"

# Iterate over each row in the DataFrame to classify products based on their descriptions.
for index, row in cleaned_data.iterrows():
    # Extract the product description from the current row.
    description = row['Description']
    # Check each category and its keywords for current description
    for category, keywords in category_keywords.items():
      # If any keyword in the current category is found in the description assign the corresponding category to this product.
        if any(keyword in description for keyword in keywords):
            cleaned_data.at[index, 'Product Category'] = category
            break

In [ ]:
cleaned_data["Product Category"].value_counts()

,count
Product Category,
"Fine Art, Collectibles, Crafts & Sewing",116629
Kitchenware,113760
Home & Living & Garden,54654
Decorations,33871
"Clothing, Shoes, Accessories, Jewelry & Beauty",15335
Lightening,12900
Toys & Games,12716
Paper Products,8336
Seasonal & Holiday,5778


In [ ]:
cleaned_data.drop('Category', axis=1, errors='ignore', inplace=True)

In [ ]:
# Calculate sales and revenue by category

# Clculate number of quantities sold for each category to identify sales_by_categories
sales_by_category = cleaned_data.groupby('Product Category')['Quantity'].sum().reset_index()
# Clculate summation of total price sold for each category to identify revenue_by_categories
revenue_by_category = cleaned_data.groupby('Product Category')['Totalprice'].sum().reset_index()

sales_by_category.columns = ['Product Category', 'Total Sales']
revenue_by_category.columns = ['Product Category', 'Total Revenue']

In [ ]:
sales_by_category

,Product Category,Total Sales
0,Appliances,2095
1,Automotive Parts & Accessories,9651
2,Baby Supplies,1880
3,Books,7857
4,"Clothing, Shoes, Accessories, Jewelry & Beauty",274487
5,Decorations,431637
6,"Fine Art, Collectibles, Crafts & Sewing",1645793
7,Home & Living & Garden,613646
8,Kitchenware,1347403
9,Lightening,228474


In [ ]:
revenue_by_category

,Product Category,Total Revenue
0,Appliances,1609.540
1,Automotive Parts & Accessories,14071.050
2,Baby Supplies,2403.160
3,Books,12068.490
4,"Clothing, Shoes, Accessories, Jewelry & Beauty",410734.980
5,Decorations,706625.280
6,"Fine Art, Collectibles, Crafts & Sewing",2725080.060
7,Home & Living & Garden,1239293.510
8,Kitchenware,2641583.270
9,Lightening,400142.720


In [ ]:
top_5_categories = revenue_by_category.sort_values(by='Total Revenue', ascending=False).head(5)

fig = px.bar(top_5_categories,
             x='Total Revenue',
             y='Product Category',
             title='Top 5 Categories by Revenue',
             color='Product Category',
             color_continuous_scale='Set1',
             orientation='h')

fig.update_layout(
    xaxis_title='Total Revenue (in millions)',
    yaxis_title='Product Category',
    title_x=0.5
)

fig.show()

##Insights gained from the revenue analysis are:
- The highest product category is "Fine Art, Collectibles, Crafts & Sewing" and comes in second and third place "Kitchenware" and "Home & Living & Garden".

# **Task6: Geographical Analysis**

## Analyze sales distribution by country and identify top revenue-generating countries.

In [ ]:
# Calculate total revenue by country
revenue_by_country = cleaned_data.groupby('Country')['Totalprice'].sum().sort_values(ascending=False).head(10).reset_index()
revenue_by_country

,Country,Totalprice
0,United Kingdom,7285024.644
1,Netherlands,285446.340
2,EIRE,265262.460
3,Germany,228678.400
4,France,208934.310
5,Australia,138453.810
6,Spain,61558.560
7,Switzerland,56443.950
8,Belgium,41196.340
9,Sweden,38367.830


In [ ]:
# A bar plot of the top 10 countries by revenue
fig = px.bar(revenue_by_country,
             x='Country',
             y='Totalprice',
             title='Top 10 Countries by Revenue',
             color='Country',
             color_continuous_scale='Blues')

fig.update_layout(
    xaxis_title='Country',
    yaxis_title='Total Revenue (in millions)',
    title_x=0.5,
    xaxis_tickangle=45,
    yaxis_tickformat=".1fM",
    yaxis_gridcolor='rgba(0,0,0,0.2)',
    yaxis_gridwidth=0.7
)

fig.show()

In [ ]:
# Percentage of sales from top 3 countries
top_3_countries_revenue = revenue_by_country.head(3).sum()
total_revenue = revenue_by_country.sum()

percentage_top_3 = (top_3_countries_revenue.Totalprice / total_revenue.Totalprice) * 100
print(f"Percentage of revenue from top 3 countries: {percentage_top_3:.2f}%")

Percentage of revenue from top 3 countries: 91.01%


# **Insigths**

The top 3 countries represent most of the revenue (more than **90%**), especially **United Kingdom**

# **Task7: Customer Behavior Analysis**

## Analyze customer behavior based on order quantity, sales, and purchasing patterns.

- Plot the distribution of order quantities.
- Create a scatter plot of Quantity vs. TotalPrice.
- Calculate and plot average daily sales throughout the week.

In [ ]:
# Explore Quantity Statistics
cleaned_data['Quantity'].describe()

,Quantity
count,392732.000000
mean,13.153718
std,181.588420
min,1.000000
25%,2.000000
50%,6.000000
75%,12.000000
max,80995.000000


In [ ]:
fig = px.histogram(cleaned_data,
                   x='Quantity',
                   nbins=50,
                   title='Distribution of Order Quantities',
                   color_discrete_sequence=['lightgreen'])

fig.update_layout(
    xaxis_title='Order Quantity',
    yaxis_title='Frequency',
    title_x=0.5,
    bargap=0.05
)

fig.update_traces(marker_line_color='black', marker_line_width=1)

fig.show()

# **Observations**

It's very clear from summary statistics and histogram that Quantity distribution is skwed and have extreme values

- The mean is 13.15, and the standard deviation is 181.59, indicating extreme outliers (as the max is 80,995).
- The median is 6, showing that most values are concentrated on the lower end.
- The 75th percentile is 12, meaning 75% of the data is below this value.

In order to make the plot more clear and readable to furture explore quantity distribution we will filter the data to focus more on main values.

In [ ]:
# Visualize Quantity Distribution after filtering data
fig = px.histogram(cleaned_data[cleaned_data['Quantity'] <= 100],
                   x='Quantity',
                   nbins=20,
                   title='Distribution of Order Quantities',
                   color_discrete_sequence=['lightgreen'])

fig.update_layout(
    xaxis_title='Order Quantity',
    yaxis_title='Frequency',
    title_x=0.5,
    bargap=0.05
)

fig.update_traces(marker_line_color='black', marker_line_width=1)

fig.show()

In [ ]:
# Scatter plot of Quantity vs Totalprice
fig = px.scatter(
    cleaned_data,
    x='Quantity',
    y='Totalprice',
    title='Quantity vs Totalprice',
)

fig.update_layout(
    xaxis_title='Quantity',
    yaxis_title='Totalprice',
)

fig.show()

In [ ]:
# Calculate and plot average daily sales throughout the week.
cleaned_data['Weekday'] = cleaned_data['InvoiceDate'].dt.day_name()
average_sales_by_weekday = cleaned_data.groupby('Weekday')['Totalprice'].mean().reindex(
    ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
)

fig = px.bar(
    average_sales_by_weekday,
    x=average_sales_by_weekday.index,
    y=average_sales_by_weekday.values,
    title='Average Daily Sales Throughout the Week',
    labels={'x': 'Day of the Week', 'y': 'Average Sales'}
)

fig.update_traces(marker_line_color='black', marker_line_width=1)
fig.update_layout(
    xaxis_title='Day of the Week',
    yaxis_title='Average Sales'
)

fig.show()

# **Insights**

- Most values of Sold Quantities ranges from 1 to 30 with 2 extreme values.
- Most of Sales happend on Friday and Tuesday, with nearly similar values on (Thursday, Wednesday, Monday), low sales on Sunday and nearly no sales on Saterday.
- Total sales increases with number of sold quantities.

# **Task8: Moving Average Forecast**

In [ ]:
# Extract daily sales by calculating total sales for each day in data
daily_sales = cleaned_data.groupby(cleaned_data['InvoiceDate'].dt.date)['Totalprice'].sum().reset_index()
daily_sales.columns = ['Date', 'TotalSales']
daily_sales = daily_sales.sort_values(by='Date')

# Calculate a 7-day Moving Average using rolling() function with window size = 7
daily_sales['7DayMovingAvg'] = daily_sales['TotalSales'].rolling(window=7).mean()

# Extract actual sales for the last 3 months to compare with the moving average
daily_sales['Date'] = pd.to_datetime(daily_sales['Date'])
last_three_months = daily_sales[daily_sales['Date'].dt.date >= (daily_sales['Date'].max() - pd.DateOffset(months=3)).date()]

last_three_months.tail(10)

,Date,TotalSales,7DayMovingAvg
295,2011-11-29,48640.57,43329.335714
296,2011-11-30,41396.79,42180.512857
297,2011-12-01,44405.37,38278.030000
298,2011-12-02,44526.10,39158.945714
299,2011-12-04,20232.00,38257.470000
300,2011-12-05,58081.09,44130.590000
301,2011-12-06,45989.66,43324.511429
302,2011-12-07,69230.60,46265.944286
303,2011-12-08,50395.96,47551.540000
304,2011-12-09,184329.66,67540.724286


In [ ]:
# Create a Plotly figure
fig = go.Figure()

# Add actual sales trace
fig.add_trace(go.Scatter(
    x=last_three_months['Date'],
    y=last_three_months['TotalSales'],
    mode='lines+markers',
    name='Actual Sales',
    line=dict(color='blue')
))

# Add moving average trace
fig.add_trace(go.Scatter(
    x=last_three_months['Date'],
    y=last_three_months['7DayMovingAvg'],
    mode='lines',
    name='7-Day Moving Average',
    line=dict(color='orange', dash='dash')
))

fig.update_layout(
    title='Daily Sales vs. 7-Day Moving Average (Last 3 Months)',
    xaxis_title='Date',
    yaxis_title='Sales',
    legend_title='Legend'
)

fig.show()

## **Insights**

- **Actual daily sales**
    - There are clear spikes in daily sales at regular intervals, indicating possible promotional events, campaigns, or specific days of increased customer activity.
    - A significant spike occurs in early December, possibly tied to a major sale event or seasonal demand.
- **The moving average**
    - The 7-day moving average (orange line) smooths out the fluctuations in daily sales, providing a clearer view of the underlying trend, shows a steady pattern with no drastic drops, suggesting consistent sales performance over time.
- **Overall Seasonality & Trend**
    - Sales appear to increase significantly as December approaches, potentially indicating the impact of holiday shopping or year-end sales.
    - Prior to December, daily sales remain relatively stable, with periodic spikes.


# **Task 9: Summary Dashboard Creation**

In [ ]:
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        "Monthly Sales Trend",
        "Top 5 Products by Revenue",
        "Customer Segment Distribution",
        "Top 5 Countries by Revenue"
    ),
    specs=[[{"type": "xy"}, {"type": "xy"}],
           [{"type": "domain"}, {"type": "xy"}]]  # types for each subplot
)


# Monthly sales trend
fig.add_trace(
    go.Scatter(x=monthly_sales['Month'], y=monthly_sales['TotalSales'], mode='lines+markers', name="Monthly Sales"),
    row=1, col=1
)


# Top 5 products by revenue
top_products_5 = top_products.head(5)
fig.add_trace(
    go.Bar(x=top_products_5['Description'], y=top_products_5['Quantity'], name="Top Products"),
    row=1, col=2
)


# Customer segment distribution
segment_distribution = customer_segments['Segment'].value_counts()

fig.add_trace(
    go.Pie(labels=segment_distribution.index, values=segment_distribution.values, name="Customer Segments"),
    row=2, col=1
)


# Top 5 countries by revenue
revenue_by_country_5 = revenue_by_country.head(5)
fig.add_trace(
    go.Bar(x=revenue_by_country_5['Country'], y=revenue_by_country_5['Totalprice'], name="Top Countries", marker_color='orange'),
    row=2, col=2
)



fig.update_layout(
    height=800, width=1000,
    title_text="Key Insights Dashboard",
    showlegend=False
)

fig.show()

# **Task 10: Optimize Data Processing**

Compare and report performance differences between using loops and vectorized operations.

In [ ]:
# Performing Top_countries_by_Revenue task using loops and Vectorization

# 1) loop
def top_countries_loop(data, n=5):
    country_revenue = {}
    for _, row in data.iterrows():
        country = row['Country']
        revenue = row['Totalprice']
        if country not in country_revenue:
            country_revenue[country] = 0
        country_revenue[country] += revenue
    sorted_revenue = sorted(country_revenue.items(), key=lambda x: x[1], reverse=True)
    return dict(sorted_revenue[:n])


start_time = time.time()
top_countries_loop_result = top_countries_loop(cleaned_data)
loop_time = time.time() - start_time


# 2) Vectorized

def top_countries_vectorized(dataa, n=5):
    return dataa.groupby('Country')['Totalprice'].sum().nlargest(n)


start_time = time.time()
top_countries_vectorized_result = top_countries_vectorized(cleaned_data)
vectorized_time = time.time() - start_time


# Compare results and performance
print(f"Loop Time: {loop_time:.5f} seconds")
print(f"Vectorized Time: {vectorized_time:.5f} seconds")
speedup = loop_time / vectorized_time
print(f"Speedup: {speedup:.2f}x")


Loop Time: 22.64490 seconds
Vectorized Time: 0.04558 seconds
Speedup: 496.78x


In [ ]:
# Plotting
fig = make_subplots(rows=1, cols=2)

# SpeedUp Comparison Plot
fig.add_trace(
    go.Bar(
        x=["Victorized","Loop"],
        y=[speedup,1],
        name="SpeedUp Comparison",
        marker_color='purple'
    ),
    row=1, col=1
)

# Time Comparison Plot
fig.add_trace(
    go.Bar(
        x=["Vectorized","Loop"],
        y=[vectorized_time,loop_time],
        name="Time Comparison",
        marker_color='Red'
    ),
    row=1, col=2
)


# Add annotation for lowest value in speed-up comparison
fig.add_annotation(
    x="Loop", y=1,
    text="Lowest value: 1x (Least efficient).",
    showarrow=True,
    arrowhead=2,
    ax=20, ay=-40,
    font=dict(size=12, color="black"),
    arrowcolor="black",
    row=1, col=1
)

# Add annotation for lowest value in time comparison
fig.add_annotation(
    x="Vectorized", y=vectorized_time,
    text=f"Lowest value: {vectorized_time:.3f} (Most efficient).",
    showarrow=True,
    arrowhead=2,
    ax=-20, ay=-40,
    font=dict(size=12, color="black"),
    arrowcolor="black",
    row=1, col=2
)


fig.update_layout(
    title="Performance Comparison: Execution Time and Speedup",
    height=600,
    legend_title="Performance Metrics",
    xaxis_title="Method",
    yaxis_title="Speedup (Loop/Vectorized)",
    xaxis2_title="Method",
    yaxis2_title="Execution Time (seconds)"
)


fig.show()

# **Insights**


- Using vectorization is way more efficient.

- While loop Time was 24.14020 seconds  Vectorized Time was 0.02962 seconds which indicates that vectorized functions is 815.01x faster than loop (DIFFERENT EVERY RUN).

# **Task 11: Report Generation**

In the following link you can find a summary report with key insights and recommendations [Link](https://docs.google.com/document/d/1en7sZBYdmLaObd7OM2ygVSmHhKkDQ3n3/edit?usp=sharing&ouid=115575227701801776440&rtpof=true&sd=true)